From e170c0e26ae352234a6161b8497c96449e85b0c7 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:37:34 -0400
Subject: [alsa] hda: Fix invalid function call in snd_hda_add_vmaster()

Message-id: <140972823573-369-patchset-jkysela@redhat.com>
Patchwork-id: 91065
O-Subject: [RHEL 7.1 PATCH 368/442] ALSA: hda - Fix invalid function call in snd_hda_add_vmaster()
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

The recent commit [6194b99d: ALSA: hda - Kill the rest of snd_print*()
usages] changed the callback map_slaves(), but one call was forgotten
to be replaced due to the cast, which leads to kernel Oops due to
invalid function.  This patch replaces it with a proper function.

Fixes: 6194b99de9f5 ('ALSA: hda - Kill the rest of snd_print*() usages')
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: e8750940ce345bf15fa0756810ae9ac491e9e352)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/hda_codec.c b/sound/pci/hda/hda_codec.c
index d91cb7f..47a6177 100644
--- a/sound/pci/hda/hda_codec.c
+++ b/sound/pci/hda/hda_codec.c
@@ -2831,6 +2831,12 @@ static int init_slave_unmute(struct hda_codec *codec,
  return put_kctl_with_value(slave, 1);
 }
 
+static int add_slave(struct hda_codec *codec,
+       void *data, struct snd_kcontrol *slave)
+{
+ return snd_ctl_add_slave(data, slave);
+}
+
 /**
  * snd_hda_add_vmaster - create a virtual master control and add slaves
  * @codec: HD-audio codec
@@ -2873,8 +2879,7 @@ int __snd_hda_add_vmaster(struct hda_codec *codec, char *name,
  if (err < 0)
   return err;
 
- err = map_slaves(codec, slaves, suffix,
-    (map_slave_func_t)snd_ctl_add_slave, kctl);
+ err = map_slaves(codec, slaves, suffix, add_slave, kctl);
  if (err < 0)
   return err;
 
-- 
1.7.1